[aoti] package weights to disk and dedup

yushangdi opened the pull request:
We package the weights and save them in `data/weights/` (`WEIGHTS_DIR`). In addition, we store a `weights_config.json` in the model folder for each model to specify which weight file corresponding to which weight name. 

Models can share weights. We dedup the weights based on their underlying storage (`tensor.untyped_storate()`). 


- Use `"aot_inductor.package_constants_on_disk": True` config to produce the `Weights` in aot_compile
- If we see `Weights` in aoti_files, we'll automatically package them to disk
- `"aot_inductor.package_constants_on_disk"` config and `"aot_inductor.package_constants_in_so"` config work independently.
- Use `load_pt2(package_path, load_weights_from_disk=True)` to load the weights from disk. `load_weights_from_disk` defaults to False.


Test Plan:
```
buck2 run @//mode/dev-nosan //caffe2/test/inductor:aot_inductor_package -- -r "test_package_shared_weights"
```

Tested with whisper at https://github.com/pytorch-labs/torchnative/pull/7

Rollback Plan:

Differential Revision: D74747190




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

yushangdi commented on torch/_inductor/utils.py:
```diff
@@ -3110,3 +3110,116 @@ def is_codegen_graph_partition_subgraph(wrapper: PythonWrapperCodegen) -> bool:
+        for weight_name, (_, properties) in weights.items():
+            weights_dict[properties.storage_ptr].add((model_name, weight_name))
+
+    return list(weights_dict.values())
```
> should this code be under export/pt2_archive?

sure I can move them there

desertfire reviewed (commented):
It probably will come as follow-up PR, but we also need to make this work when package_cpp_only.

yushangdi commented:
> It probably will come as follow-up PR, but we also need to make this work when package_cpp_only.

@desertfire The currently implementation works with `package_cpp_only` already if you load the `.pt2` from python API. For example, if you remove the `@skipif(...package_cpp_only,...` on my unit tests, they'll still pass. 

What's left to do is 

1) clean up the package cpp only route by removing the following steps

- package duplicated weights in codecache.py. If we package weights to disk, we don't need to package weights to serialized_weights.bin or consts_o

- avoid loading the weights from consts_o or serialized_weights.bin. The weights will be overwritten by the weights loaded from disk anyway


2) the currently implementation only works if you load .pt2 from python. If we want it to work when we load from .pt2 in cpp as well, we need to do the following

- implement the equivalent of `load_pt2` in C++, which means we need a function to load all models in one call, and manage the weight sharing. 
- weights are currently saved with torch.save, which cannot be easily loaded in C++. We need to save the serialized bytes (maybe we can re-use _to_bytes()?). 

cc @angelayi ^^ summary of our discussion

yushangdi committed:
[aoti] package weights to disk and dedup (#155241)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/155241

We package the weights and save them in `data/weights/` (`WEIGHTS_DIR`). In addition, we store a `weights_config.json` in the model folder for each model to specify which weight file corresponding to which weight name.

Models can share weights. We dedup the weights based on their underlying storage (`tensor.untyped_storate()`).

- Use `"aot_inductor.package_constants_on_disk": True` config to produce the `Weights` in aot_compile
- If we see `Weights` in aoti_files, we'll automatically package them to disk
- `"aot_inductor.package_constants_on_disk"` config and `"aot_inductor.package_constants_in_so"` config work independently.
- Use `load_pt2(package_path, load_weights_from_disk=True)` to load the weights from disk. `load_weights_from_disk` defaults to False.

Test Plan:
```
buck2 run @//mode/dev-nosan //caffe2/test/inductor:aot_inductor_package -- -r "test_package_shared_weights"

buck2 run @//mode/dev-nosan //caffe2/test/inductor:aot_inductor_package -- -r "test_package_weights_on_disk_nested_module"

buck2 run @//mode/dev-nosan //caffe2/test/inductor:test_aot_inductor -- -r test_weight_on_disk_legacy
```

Rollback Plan:

Reviewed By: desertfire

Differential Revision: D74747190
